### DISCUSSION TITLE
Documentation for github.event objects

### DISCUSSION THREAD

**Discussion by MrSkwiggs** [community/community#27058] [Actions]
<p>Is there a way to see the actual documentation / specs for the <code>github</code> object we can access when setting up workflows? I tried to echo <code>github.event</code> for instance but that just prints <code>object</code>…</p>
<p>My end goal is to have a workflow run only on PR merges or manual triggers, so I have the following configuration:</p>
<pre><code class="lang-yaml">on:
  workflow_dispatch:
  pull_request:
    types: [closed]
</code></pre>
<p>and in my job, I do indeed need to check if the PR is merged</p>
<pre><code class="lang-yaml">my-run:
    runs-on: [ubuntu-latest]
    if: ${{ github.event.pull_request.merged }}
    # more steps
</code></pre>
<p>but that means that if I trigger a manual workflow, the <code>if</code> clause prevents it from ever running… How can tell my <code>if</code> that I also want to run on <code>workflow_dispatch</code>? I tried to add <code>|| github.event.workflow_dispatch</code> in there but that just doesn’t do anything… And because I can’t find exactly what is included in <code>github.event</code>, I don’t know what to try.</p>
<p>So my question is two-fold:</p>
<ul>
<li>Where can I find detailed docs for all of the <code>github.event</code> objects &amp; their sub-objects?</li>
<li>How can I have a workflow run on either a merged pr or a manual workflow_dispatch?</li>
</ul>
<p>For reference, I did read through <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="noopener nofollow ugc">https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#github-context</a> but that only covers <em>some</em> of the docs</p>
<p><strong>UPDATE</strong><br>
So searching around I managed to find a way to echo any object by using <code>toJson(&lt;object&gt;)</code>. But for <code>workflow_dispatch</code> I couldn’t find any relevant information :confused:</p>

**Comment by airtower-luna** [comment]
<p>The best documentation I’m aware of is here: <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows">Events that trigger workflows</a></p>
<p>For most event types there’s a link to the webhook event payload documentation in the section on that event type on that page, e.g. <a href="https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/webhook-events-and-payloads#workflow_dispatch">here is the one for <code>workflow_dispatch</code></a>. I agree that is kind of complicated to find.</p>

**Comment by MrSkwiggs** [comment]
<p>Thanks for the response. Unfortunately I had already seen this page too and there isn’t much there either in terms of verifying that the event is indeed for <code>workflow_dispatch</code>.</p>

**Comment by MrSkwiggs** [comment, marked as answer]
<p>I did manage to find how to do so by first adding a step to print out the <code>github</code> object.</p>
<pre><code class="lang-yaml">steps:
  - name: Echo github obj
    run: echo ${{ toJson(github) }}
</code></pre>
<p>This allowed me to see that the github object contains an <code>event_name</code> key, which, in the case of a <code>workflow_dispatch</code> event, contains <code>workflow_dispatch</code>.</p>
<p>It was then very easy to modify my <code>if</code> clause to also check for this key.</p>
<pre><code class="lang-yaml">my-job:
  runs-on: [ubuntu-latest]
  if: github.event.pull_request.merged || github.event_name == 'workflow_dispatch'
  steps:
    # more steps
</code></pre>
<p><strong>note</strong>: somehow using double quotes around <code>"workflow_dispatch"</code> in the <code>if</code> clause raises an error (Unknown identifier), so make sure to use single quotes</p>

**Reply by LucasMMota** [reply]
GitHub should document this better for us. It's hard to understand how things work

**Comment by izambl** [comment]
<p>Using <code>echo</code> threw some errors for me<br>
I used this one to avoid errors:</p>
<pre><code class="lang-auto">      - name: Check object
        run: |
          cat &lt;&lt; OBJECT
          ${{ toJson(github) }}
          OBJECT
</code></pre>

**Reply by wesleyscholl** [reply]
I also had issues running: `run: echo ${{ toJson(github) }}`

```
- name: Check object
        run: |
          cat << OBJECT
          ${{ toJson(github) }}
          OBJECT
   ```
   ^^^ Worked for me - thanks! And I also agree that this should be better documented. Bits and pieces are fragmented over many GitHub Docs pages.

**Reply by believelody** [reply]
For those who still search documentation about payload, here is a link which can help y'all:
https://docs.github.com/en/webhooks/webhook-events-and-payloads

**Reply by chuckwondo** [reply]
FYI, you need single quotes around what you're echoing: `echo '${{ toJson(github) }}'`

You cannot use double quotes because the value returned by the call to `toJson` contains double quotes itself. Further, you must quote the result _somehow_ because the result also contains newlines, so using single quotes does the trick in this situation.
